scrub compaction: count validation errors and return status over the rest api

Deexie opened the pull request:
Currently, scrub returns to user the number indicating operation
result as follows:
- 1 when the operation was aborted;
- 3 in validate and segregate modes when validation errors were found
  (and in segregate mode - fixed);
- 0 if operation ended successfully.

To achieve so, if an operation was aborted in abort mode, then
the exception is propagated to storage_service.cc. Also the number
of validation errors for current scrub is gathered and summed
from each shard there.

The number of validation errors is counted and registered in metrics.
Metrics provide common counters for all scrub operation within
a compaction manager, though. Thus, to check the exact number
of validation errors, the comparison of counter value before and after
scrub operation needs to be done.

Deexie commented:
Currently it does not compile, since seastar submodule hasn't been updated in scylla repository. Should I provide a workaround for it? (locally with the new commit cherry-picked it does compile)

Deexie commented:
@bhalevy @amnonh 
Draft PR for [#9289](https://github.com/scylladb/scylla/issues/9289) issue with the yesterday's decisions applied.
Review please.

raphaelsc commented:
> Currently it does not compile, since seastar submodule hasn't been updated in scylla repository. Should I provide a workaround for it? (locally with the new seastar commit cherry-picked it does compile)

Does 'git submodule sync; git submodule update --init --recursive' fix it?

bhalevy commented on api/storage_service.cc:
```diff
@@ -1260,6 +1260,13 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
     });
 }
 
+enum class scrub_return {
```
s/scrub_return/scrub_status/

bhalevy commented on api/storage_service.cc:
```diff
@@ -1403,16 +1410,23 @@ void set_snapshot(http_context& ctx, routes& r, sharded<db::snapshot_ctl>& snap_
+        }).then_wrapped([] (auto f) {
+            if (f.failed()) {
+                return make_ready_future<json::json_return_type>(static_cast<int>(scrub_return::aborted));
+            } else if (f.get().validation_errors == 0) {
```
`f.get0()` is more appropriate here.

denesb replied:
`get0()` is deprecated, we no longer support variadic futures.

Deexie replied:
is get() ok then?

bhalevy commented on compaction/compaction.cc:
```diff
@@ -1373,6 +1402,12 @@ class scrub_compaction final : public regular_compaction {
             , _reader(std::move(underlying))
             , _validator(*_schema)
         { }
+        uint64_t get_validation_errors() override {
```
method should be specified as `const noexcept`

bhalevy commented on readers/flat_mutation_reader_v2.hh:
```diff
@@ -174,6 +174,14 @@ public:
         void set_close_required() { _close_required = true; }
         static constexpr size_t default_max_buffer_size_in_bytes() { return 8 * 1024; }
 
+        virtual uint64_t get_validation_errors() {
```
Hmm, the generic reader should not be changed for that.
Let's find another way to propagate these stats.

Deexie replied:
would need more time for this one. All of my current ideas have significant drawbacks:

1. As like in current solution, it would require changes in generic reader. 
Class reader in which we are able to count the number of validation errors is an implementation of an abstract layer flat_mutation_reader_v2, thus direct access to its fields is impossible without changes in abstract layer (like in current solution).

2. We could pass a shared_ptr to scrub_compaction members and let reader update them directly (which I don't think is elegant), but make_sstable_reader() is const. Thus const_cast would need to be used or make_sstable_reader() would need to loose its const qualifier. Neither seems appropriate.

3. We could pass a reference (or pointer set to nullptr by default) to the counter through reader's methods, though again it would require changes in generic reader.

4. Playing with dynamic_cast and checking whether we deal with reader object would need to access private nested class reader, require changes in generic reader, and probably cause dependency problems.

I would need to dive deeper in the code, since currently I do not see any good solution. I would appreciate any tips.

Deexie replied:
I did so, but then when I try to reach the counters from `compaction_scrub` I have to go through the abstract layer. That makes me write getters there and therefore also in `flat_mutation_reader_v2::impl`.

bhalevy replied:
See if you can store the counters in `compaction_data`.
@raphaelsc what do you think?

Deexie replied:
I am not sure what it would change. `compaction_data` cannot be reached from the `flat_mutation_reader_v2` or `flat_mutation_reader_v2::impl`. Do you mean I should keep the counters there instead of `compaction_result`?

Also I have recently pushed the version that does not affect generic reader. Could you review it, please?

denesb replied:
Adding fields to the abstract interface to accommodate a single user is a no-go in the vast majority of cases.
I think (2) is the way to go. You can make your counter struct `mutable` and hence be able to pass non-const reference to the reader constructor (use const-cast if needed).

bhalevy commented on compaction/compaction_manager.hh:
```diff
@@ -91,7 +91,20 @@ public:
+                validation_errors += r.validation_errors;
+                return *this;
+            }
+            friend compaction_task_stats operator+(compaction_task_stats l, const compaction_task_stats& r) {
```
typically operator plus receives `const compaction_task_stats&` for the lhs side, no?
it's clearer if it copies the left hand side arg explicitly here.

bhalevy commented on compaction/compaction_manager.hh:
```diff
@@ -284,11 +297,14 @@ private:
     std::unique_ptr<strategy_control> _strategy_control;
 private:
-    future<> perform_task(shared_ptr<task>);
+    future<task::compaction_task_stats> perform_task(shared_ptr<task>);
```
If most tasks have no stats to return, maybe this should be `future<std::optional<task::compaction_task_stats>>>`
and most tasks would return a disengaged optional rather than the default-constructed task stats.

bhalevy commented on test/boost/sstable_compaction_test.cc:
```diff
@@ -2616,7 +2619,9 @@ SEASTAR_TEST_CASE(sstable_scrub_segregate_mode_test) {
-            compaction_manager.perform_sstable_scrub(table->as_table_state(), opts).get();
+            try {
+                compaction_manager.perform_sstable_scrub(table->as_table_state(), opts).get();
+            } catch (sstables::compaction_aborted_exception& e) {}
```
If the intention is to swallow this exception in must be more visible, with a comment explaining and documenting it, and in a separate patch of course.
e.g.
```
    try {
        compaction_manager.perform_sstable_scrub(table->as_table_state(), opts).get();
    } catch (sstables::compaction_aborted_exception& e) {
        // Ignored, since x y z.
    }
```

Deexie replied:
After the changes the exception even **should** be caught here. So I just check whether it was swallowed, analogically like in other test cases.

bhalevy replied:
I don't understand

Deexie replied:
In this case `perform_sstable_scrub` is performed on invalid sstable in abort mode, so an exception is expected. Thus I check whether it was caught.

bhalevy commented on test/rest_api/test_storage_service.py:
```diff
@@ -139,6 +139,14 @@ def test_storage_service_keyspace_scrub(cql, this_dc, rest_api):
+    with new_test_keyspace(cql, f"WITH REPLICATION = {{ 'class' : 'NetworkTopologyStrategy', '{this_dc}' : 1 }}") as keyspace:
+        with new_test_table(cql, keyspace, "a int, PRIMARY KEY (a)") as t0:
+            with scylla_inject_error(rest_api, "rest_api_keyspace_scrub_abort"):
+                cql.execute(f"INSERT INTO {t0} (a) VALUES (42)")
```
Don't we need to flush the table first? Otherwise there will be no sstables to scrub.

bhalevy commented:
@Deexie general comment: please prepare the patchset as a series of targeted patches, each one doing a single self-contained change because currently the single big patch contains everything and is hard to review.

bhalevy commented:
Also, the subject should better describe what this series is doing overall.
E.g. "scrub compaction: count validation errors and fixes and return status over the rest api"

denesb commented on api/storage_service.cc:
```diff
@@ -1260,6 +1260,13 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
+enum class scrub_return {
+    successful = 0,
+    aborted,
+    unable_to_cancel,
```
What does unable to cancel mean?

bhalevy replied:
I'm not sure I fully understand what they meant by it in C*.  Maybe if there was an attempt to stop the compaction but it finished anyway.

We don't have to ever return it.

bhalevy replied:
No, it can be present, but let's add a comment that it is not used in scylla.

denesb commented:
I will wait with the review until the patch is split.

bhalevy commented:
On Wed, Jul 20, 2022, 05:52 Botond Dénes ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In api/storage_service.cc
> <https://github.com/scylladb/scylla/pull/11074#discussion_r925114847>:
>
> >                      auto& cm = db.get_compaction_manager();
>                      auto& cf = db.find_column_family(keyspace, cfname);
>                      return cm.perform_sstable_scrub(cf.as_table_state(), opts);
> -                });
> -            });
> -        }).then([]{
> -            return make_ready_future<json::json_return_type>(0);
> +                }, compaction_manager::task::compaction_task_stats{}, std::plus<compaction_manager::task::compaction_task_stats>());
> +            }, compaction_manager::task::compaction_task_stats{}, std::plus<compaction_manager::task::compaction_task_stats>());
> +        }).then_wrapped([] (auto f) {
> +            if (f.failed()) {
> +                return make_ready_future<json::json_return_type>(static_cast<int>(scrub_return::aborted));
> +            } else if (f.get().validation_errors == 0) {
>
> get0() is deprecated, we no longer support variadic futures.
>

Good point. I didn't think of it as related to varaiadic futures...

> Message ID: ***@***.***>
>

bhalevy commented on api/storage_service.cc:
```diff
@@ -1403,16 +1411,33 @@ void set_snapshot(http_context& ctx, routes& r, sharded<db::snapshot_ctl>& snap_
+
+        const auto& reduce_validation_errors = [] (const compaction_manager::task::validation_errors_stats& lhs, const compaction_manager::task::validation_errors_stats& rhs) {
+            compaction_manager::task::compaction_task_stats cts{};
+            if (lhs) {
```
How can it be null?

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -277,16 +279,17 @@ compaction_manager::task::task(compaction_manager& mgr, compaction::table_state*
 {}
 
-future<> compaction_manager::perform_task(shared_ptr<compaction_manager::task> task) {
+future<compaction_manager::task::validation_errors_stats> compaction_manager::perform_task(shared_ptr<compaction_manager::task> task) {
```
I hope that by storing the validation counters in `compaction_data` (similar to `total_partitions` and `total_keys_written`), we can copy them to `compaction_result` in `compaction::finish()`, and we can return it here.
It makes more sense to return a generic `compaction_result`, then scrub specific stats.

Deexie replied:
If we returned `compaction_result` from each function up to `perform_task`, then we would need to go through `do_run`:
https://github.com/scylladb/scylla/blob/d566755234db2f7754a7ca0fa49e390ab1319836/compaction/compaction_manager.cc#L1262-L1272
and thus we would need to gather the returned value from each `validate_sstable`. Do we want to join `new_sstables` vectors and pass it through many functions, while we do not really use it anyhow?

I was thinking about creating a `compaction_counters` struct with `fixed_validation_errors` and `validation_errors` (and possibly some others), making the field of that type inside `compaction_result`. That would be effectively equal to renaming `compaction_task_stats` (and `validation_errors_stats`), though.

What do you think about it?

bhalevy replied:
Sounds good, but I'll need to see the patch to understand it better.

bhalevy commented on compaction/compaction.cc:
```diff
@@ -1707,15 +1759,18 @@ static future<compaction_result> scrub_sstables_validate_mode(sstables::compacti
 
     clogger.info("Finished scrubbing in validate mode {} - sstable(s) are {}", sstables_list_msg, valid ? "valid" : "invalid");
 
+    uint64_t validation_errors = 0; // TODO Deexie  I am not sure if this should be counted as 1 or for every sstable (?)
```
per sstable is good.

bhalevy commented on compaction/compaction.cc:
```diff
@@ -27,6 +27,7 @@
 #include <seastar/core/coroutine.hh>
 #include <seastar/util/closeable.hh>
 
+#include "seastar/core/shared_ptr.hh"
```
nit: please stick to the block of seastar includes above

bhalevy commented on compaction/compaction.cc:
```diff
@@ -614,6 +617,10 @@ class compaction {
     // Default range sstable reader that will only return mutation that belongs to current shard.
     virtual flat_mutation_reader_v2 make_sstable_reader() const = 0;
 
+    virtual flat_mutation_reader_v2 make_sstable_reader(compaction_task_stats_ptr) const {
```
please FIXME comment in this patch ("scrub compaction: register validation errors and fixes in metrics")
with the intent to use the compaction_task_stats_ptr,
since it's unused at this point.

bhalevy commented on compaction/compaction.cc:
```diff
@@ -1242,15 +1251,29 @@ class scrub_compaction final : public regular_compaction {
 
     private:
+        void register_validation_error() {
+            if (_scrub_mode == compaction_type_options::scrub::mode::segregate && _cts) {
```
Can _cts be disengaged in `scrub::mode::segregate`?
If not, there's no point in testing that.

bhalevy replied:
Alternatively, you can just test `if (_cts)` and count without having to rely on the scrub::mode.
It's better than sprinkling the scrub logic all over the place.

bhalevy replied:
... and I think you meant `scrub::mode::validate` here, no? :)

bhalevy commented on compaction/compaction.cc:
```diff
@@ -1242,15 +1251,29 @@ class scrub_compaction final : public regular_compaction {
         flat_mutation_reader_v2 _reader;
         mutation_fragment_stream_validator _validator;
         bool _skip_to_next_partition = false;
+        std::optional<compaction_task_stats_ptr> _cts;
```
Since `compaction_task_stats_ptr` is a `lw_shared_ptr` it doesn't need to be optional.
It's default-constructed to have no payload..

bhalevy commented on compaction/compaction.cc:
```diff
@@ -1242,15 +1251,29 @@ class scrub_compaction final : public regular_compaction {
+        }
+
+        void register_validation_error_fix() {
+            if (_scrub_mode == compaction_type_options::scrub::mode::segregate && _cts) {
```
same here

bhalevy commented on compaction/compaction.cc:
```diff
@@ -1367,11 +1398,12 @@ class scrub_compaction final : public regular_compaction {
 
     public:
-        reader(flat_mutation_reader_v2 underlying, compaction_type_options::scrub::mode scrub_mode)
+        reader(flat_mutation_reader_v2 underlying, compaction_type_options::scrub::mode scrub_mode, std::optional<compaction_task_stats_ptr> cts = std::nullopt)
```
ditto, optional is not really needed with a lw_shared_ptr

bhalevy commented on compaction/compaction.cc:
```diff
@@ -1587,6 +1637,7 @@ future<compaction_result> compaction::run(std::unique_ptr<compaction> c) {
             throw;
         }
 
+
```
nit: extraneous empty line

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -695,6 +697,10 @@ future<> compaction_manager::update_throughput(uint32_t value_mbs) {
 
+    sm::label mode_label("mode");
+    const auto segregate_mode_label_instance = mode_label("segregate");
+    const auto validate_mode_label_instance = mode_label("validate");
```
let's separate the metrics part into a patch of its own.
there's already a lot going on in this patch without that.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1133,17 +1148,20 @@ class compaction_manager::rewrite_sstables_compaction_task : public compaction_m
         while (!_sstables.empty() && can_proceed()) {
             auto sst = consume_sstable();
-            co_await rewrite_sstable(std::move(sst));
+            co_await rewrite_sstable(std::move(sst), cts);
```
it'd be cleaner if rewrite_sstable will just return the `compaction_result`
and the accumulation into `compaction_task_stats` will be done here, by the caller.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1224,14 +1244,17 @@ class compaction_manager::validate_sstables_compaction_task : public compaction_
         while (!_sstables.empty() && can_proceed()) {
             auto sst = consume_sstable();
-            co_await validate_sstable(std::move(sst));
+            co_await validate_sstable(std::move(sst), cts);
```
same here, like rewrite_sstable above

bhalevy commented on compaction/compaction_manager.hh:
```diff
@@ -91,7 +91,24 @@ public:
+                compaction_task_stats tmp{
+                    .fixed_validation_errors = l.fixed_validation_errors,
+                    .validation_errors = l.validation_errors,
+                };
```
`auto tmp = l;` would be simpler.
just engage the default copy constructor.

bhalevy commented on compaction/compaction_manager.hh:
```diff
@@ -284,11 +297,14 @@ private:
     std::optional<utils::observer<uint32_t>> _throughput_option_observer;
+    uint64_t validation_errors_segregate = 0;
+    uint64_t validation_errors_validate = 0;
+    uint64_t fixed_validation_errors = 0;
```
As I said above, let's split the metrics part into a separate patch.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -277,16 +279,17 @@ compaction_manager::task::task(compaction_manager& mgr, compaction::table_state*
 
     try {
-        co_await task->run();
+        const auto &res = co_await task->run();
```
nit: `auto res =` is simpler and more visible that you're storing the result in the automatic variable.
Then co_return res will move it (using copy elision)

`auto&& res =` is also good as a universal reference.  it might save the move here.

In any case, in our coding style, the `&` sticks to `auto`, so what you wrote should be `const auto& res =`

amnonh commented on compaction/compaction_manager.cc:
```diff
@@ -710,6 +716,15 @@ void compaction_manager::register_metrics() {
                        sm::description("Holds the sum of compaction backlog for all tables in the system.")),
         sm::make_gauge("normalized_backlog", [this] { return _last_backlog / _available_memory; },
                        sm::description("Holds the sum of normalized compaction backlog for all tables in the system. Backlog is normalized by dividing backlog by shard's available memory.")),
+        sm::make_counter("validation_errors", [this] { return validation_errors_validate; },
```
I know it's confusing, but all the description for a specific metric name should be the same

denesb commented:
When pushing a new version, please add here in a comment a changelog of what changed since the last version, so reviewers can focus on the changes only, instead of having to re-review everything from scratch.

Deexie commented:
- splitted commit with counting and registering counters in metrics
- changed the descriptions of counters so that the ones named the same have the same description
- const auto& res -> auto&& res
- added compaction_counters inside compaction_result and returned it upstream

denesb commented on compaction/compaction.cc:
```diff
@@ -1242,15 +1251,25 @@ class scrub_compaction final : public regular_compaction {
         void maybe_abort_scrub() {
             if (_scrub_mode == compaction_type_options::scrub::mode::abort) {
                 throw compaction_aborted_exception(_schema->ks_name(), _schema->cf_name(), "scrub compaction found invalid data");
             }
```
I think this would be the best place to increment error metrics. You can bump `validation_errors` before the if and `fixed_validation_errors` after the if (if we don't abort, we fix the error). BTW why count fixes too? When we don't abort, we fix all errors we find. Don't really see the point in two separate metrics, they will always have the same value.

denesb commented on compaction/compaction.cc:
```diff
@@ -1707,15 +1755,18 @@ static future<compaction_result> scrub_sstables_validate_mode(sstables::compacti
     if (!valid) {
         for (auto& sst : *sstables->all()) {
             co_await sst->move_to_quarantine();
+            ++validation_errors;
```
Why do we update this counter here?

Deexie replied:
In validate mode we do not enter any of on_invalid_* functions, so the errors should be counted somewhere else. This seems to be reasonable.

denesb replied:
But this will be increase the counter once for each sstable, whereas the validation can find many more in each sstable.

denesb commented on compaction/compaction.cc:
```diff
@@ -1707,15 +1755,18 @@ static future<compaction_result> scrub_sstables_validate_mode(sstables::compacti
     co_return compaction_result {
         .new_sstables = {},
         .ended_at = db_clock::now(),
+        .validation_errors = validation_errors,
```
Shouldn't we take this from `cts` above?

Deexie replied:
I don't understand. Which `cts` do you mean?

denesb replied:
`auto cts = seastar::make_lw_shared<compaction_manager::task::compaction_task_stats>();`

Deexie replied:
It's beyond the scope of that function though.

denesb commented on compaction/compaction.cc:
```diff
@@ -1576,7 +1623,8 @@ class resharding_compaction final : public compaction {
     return seastar::async([c = std::move(c)] () mutable {
         c->setup().get();
-        auto consumer = c->consume();
+        auto cts = seastar::make_lw_shared<compaction_manager::task::compaction_task_stats>();
```
Does this have to be shared pointer?

I think `compaction` should create this and store it as a member. It is less noise.

denesb commented on compaction/compaction_manager.hh:
```diff
@@ -301,6 +301,9 @@ private:
     serialized_action _update_compaction_static_shares_action;
     utils::observer<float> _compaction_static_shares_observer;
+    uint64_t validation_errors_segregate = 0;
+    uint64_t validation_errors_validate = 0;
```
I don't think it is interesting to separate the two. These are very rare, manually started operations.

denesb replied:
I think just a general `validation_error`, that you increase for both `scrub` and `validatate` compaction is sufficient. These two operations will never run in parallel, so the operator will always be able to tell where the validation errors are coming from.

denesb commented on test/boost/sstable_compaction_test.cc:
```diff
@@ -2519,7 +2519,14 @@ SEASTAR_TEST_CASE(sstable_scrub_skip_mode_test) {
+            try {
+                compaction_manager.perform_sstable_scrub(table->as_table_state(), opts).get();
+            }
+            catch (sstables::compaction_aborted_exception& e) {
```
catch should continue on the same line as `}`

denesb reviewed (changes requested):
I think it is pointless to track `segregate` and `validation` as well as `errors` and `fixes` metrics separately.
We can have just one external `validation_errors` metrics and no need to track `fixes` at all, they are always the same as `errors`.

Deexie commented:
- got rid of fixes counting and metric
- blended counters (and metrics) for validation and segregation mode together
- moved errors counting to `maybe_abort_scrub()`
- counted distinct validation error in each sstable in validate mode
- } \\ catch -> } catch

Deexie commented on compaction/compaction_manager.hh:
```diff
@@ -91,7 +91,19 @@ public:
         };
         static std::string_view to_string(state);
-
+        struct compaction_task_stats {
```
Do we still want to keep that struct (for documentation/future extension) or just single uint64_t?

bhalevy replied:
I think so.  It does't have any performance penalty and it makes the usage clearer.
Plus, it will help if and when we'd want to extend it.

Deexie commented on compaction/compaction_manager.cc:
```diff
@@ -1133,17 +1135,19 @@ class compaction_manager::rewrite_sstables_compaction_task : public compaction_m
             auto sst = consume_sstable();
-            co_await rewrite_sstable(std::move(sst));
+            const auto& res = co_await rewrite_sstable(std::move(sst));
+            cts.validation_errors += res.validation_errors;
```
Should FIXME comment be added here? Since currently `cts` is not used anywhere (it is in the next commit).

bhalevy replied:
yes, please

Deexie commented on compaction/compaction_manager.cc:
```diff
@@ -1144,6 +1146,7 @@ class compaction_manager::rewrite_sstables_compaction_task : public compaction_m
             const auto& res = co_await rewrite_sstable(std::move(sst));
             cts.validation_errors += res.validation_errors;
         }
+        _cm.validation_errors += cts.validation_errors;   // TODO Deexie  maybe it should be done after every rewrite_sstable to provide gradual growth (?)
```
Is it ok if `_cm.validation_errors` is updated here and not in every loop iteration?

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1132,18 +1151,24 @@ class compaction_manager::rewrite_sstables_compaction_task : public compaction_m
+            const auto& res = co_await rewrite_sstable(std::move(sst));
+            counters += res.counters;
         }
+        _cm.validation_errors += counters.validation_errors;  // TODO Deexie  maybe it should be done after every rewrite_sstable to provide gradual growth (?)
```
It would be better to update `_cm.validation_errors` in the inner loop, per sstable as the comment says.

Deexie commented:
- fixed mode for which counters are modified
- added FIXME comment
- counter value is updated after each sstable operation

bhalevy commented on compaction/compaction_manager.hh:
```diff
@@ -299,6 +299,7 @@ private:
     utils::updateable_value<float> _static_shares;
     serialized_action _update_compaction_static_shares_action;
     utils::observer<float> _compaction_static_shares_observer;
+    uint64_t validation_errors = 0;
```
Since this is a private member out convention is to start it with `_`, i.e. `_validation_errors`.

bhalevy commented on api/storage_service.cc:
```diff
@@ -1403,16 +1411,29 @@ void set_snapshot(http_context& ctx, routes& r, sharded<db::snapshot_ctl>& snap_
-        return f.then([&ctx, keyspace, column_families, opts] {
-            return ctx.db.invoke_on_all([=] (replica::database& db) {
-                return do_for_each(column_families, [=, &db](sstring cfname) {
+
```
nit: extra empty line

bhalevy commented on compaction/compaction.hh:
```diff
@@ -85,12 +85,29 @@ struct compaction_data {
+
+struct compaction_result {
+    std::vector<sstables::shared_sstable> new_sstables;
+    compaction_counters counters;
```
let's do this refactoring in a separate patch please so it'll stand out as a standalone change

bhalevy commented on compaction/compaction_manager.hh:
```diff
@@ -40,6 +40,7 @@ class compacting_sstable_registration;
 // behalf of existing tables.
 class compaction_manager {
 public:
+    using compaction_result_counters = std::optional<sstables::compaction_counters>;
```
I have doubts about this shortname.
It lets a reader think there is yet another struct holding counters that is different than `compaction_counters` itself.

In other places we use a convention for a `_opt` suffix for optional typedefs, so in this case it would be
`using compaction_counters_opt = std::optional<sstables::compaction_counters>;`
This better describes the type.

For example, I didn't figure out why in your map_reduce implementation you instatiated an optional for the initial value, until I got here...

bhalevy commented on api/storage_service.cc:
```diff
@@ -1428,8 +1428,10 @@ void set_snapshot(http_context& ctx, routes& r, sharded<db::snapshot_ctl>& snap_
+        }).then_wrapped([] (auto f) {
+            if (f.failed()) {
+                return make_ready_future<json::json_return_type>(static_cast<int>(scrub_status::aborted));
+            } else if (f.get().value().validation_errors == 0) {
```
`f.get()->validation_errors` is simpler.

bhalevy replied:
and I suggest to simplify the if/then ladder by swapping the conditional terms, like this:
```c++
        } else if (f.get()->validation_errors) {
            return make_ready_future<json::json_return_type>(static_cast<int>(scrub_status::validation_errors));
        } else {
                return make_ready_future<json::json_return_type>(static_cast<int>(scrub_status::successful));
        }
````

This makes the relationship between the validation_errors in the result and the validation_errors status clear and direct.
(and it looks like it needs to be done in the earlier patch that introduced this order)

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -873,14 +873,16 @@ inline bool compaction_manager::task::can_proceed(throw_if_stopping do_throw_if_
         cmlog.error("{}: {}: stopping", *this, e.what());
         _cm._stats.errors++;
+        if (throw_on_abort)
+            throw;
```
coding style nit: we always use curly braces to prevent accidental bugs due to misleading indentation.
```c++
        if (throw_on_abort) {
            throw;
        }
```

bhalevy commented on test/cql-pytest/util.py:
```diff
@@ -248,3 +250,20 @@ def local_process_id(cql):
+        yield
+    finally:
+        print("Disabling error injection", err)
+        response = requests.delete(f'{url}/v2/error_injection/injection/{err}')
```
This was copied from alternator/utils.py right?
If so, a preliminary patch moving it here would be appreciated.
ok with you @nyh?

bhalevy replied:
We would like to avoid proliferation of helper utility variants in each test utils "library"....

Deexie replied:
Yes, it was.
Should I delete it from test/alternator/util.py then? That would cause a problem with name shadowing I think. Am I supposed to rename one of util.py files? Or create a new one in test/cql-pytest?

bhalevy replied:
It would be best if the alternator test can import it from cql-tests/util.py
but I'm not sure about the exact mechanics.
IIRC we play with the python path in order to import util.py from another directory, so if both alternator and cql-tests have util.py in them that could be a problem.
For the rest_api test suite, we use rest_util.py so maybe we can simply rename alternator/util.py to alternator_util.py so the tests can import both this file and the cql-tests/util.py that can hold the generic library functions.

Does that make sense @nyh ?

bhalevy commented on test/boost/sstable_compaction_test.cc:
```diff
@@ -2519,7 +2519,13 @@ SEASTAR_TEST_CASE(sstable_scrub_skip_mode_test) {
+            } catch (sstables::compaction_aborted_exception& e) {
+                swallowed = true;
+            }
+            BOOST_REQUIRE(swallowed);
```
`BOOST_REQUIRE_THROW` is your friend here :)
(and calling the requirement `swallowed` is a bit misleading since it's swallowed by the test, not by the code you're testing, and BOOST_REQUIRE better check the code-under-test semantics, not what the test did...)

bhalevy commented on compaction/compaction.cc:
```diff
@@ -1751,6 +1752,10 @@ compact_sstables(sstables::compaction_descriptor descriptor, compaction_data& cd
+    if (descriptor.options.type() == compaction_type::Scrub
+            && std::get<compaction_type_options::scrub>(descriptor.options.options()).operation_mode == compaction_type_options::scrub::mode::abort) {
+                utils::get_local_injector().inject("rest_api_keyspace_scrub_abort", [&table_s] { throw compaction_aborted_exception(table_s.schema()->ks_name(), table_s.schema()->cf_name(), "scrub compaction found invalid data"); });
+    }
```
This place is wrong for injecting the error since normally there is no reason to abort here.
Let's inject inline in the code itself where we check if scrub needs to abort (and the error injection should simulate a non-recoverable error for scrub).
Then it can apply also to other scrub modes, not only `scrub::mode::abort`.

bhalevy commented on api/storage_service.cc:
```diff
@@ -1428,8 +1428,10 @@ void set_snapshot(http_context& ctx, routes& r, sharded<db::snapshot_ctl>& snap_
-            if (f.value().validation_errors == 0) {
+        }).then_wrapped([] (auto f) {
+            if (f.failed()) {
+                return make_ready_future<json::json_return_type>(static_cast<int>(scrub_status::aborted));
```
Let's keep this status code only for `compaction_aborted_exception` so if scrub failed for any other reason (e.g. bad_alloc) the failure would still be represented as an error to execute the api call - since the user doesn't know in this case if scrub even ran. Maybe it failed early and didn't even get to scrub any sstable - hence the error is inconclusive.

you can use something like:
```c++
    if (f.failed()) {
        auto ex = f.get_exception()
        if (try_catch<sstables::compaction_aborted_exception>(ex)) {
                return make_ready_future<json::json_return_type>(static_cast<int>(scrub_status::aborted));
        } else {
                return make_exception_future<json::json_return_type>(std::move(ex));
        }
    }
```

That said, note that the code in this patch drops the exception without officially handling it, i.e. it never calls f.get_exception().
We have a sanity check in seastar to report unhandled exception in future so when the future is destroyed while hold an exception seastar will warn about it, JSYK :)

Deexie commented:
- validation_errors -> _validation_errors
- compaction_result_counters -> compaction_counters_opt
- moved compaction_counters extraction to a separate patch
- f.get().value().validation_errors -> f.get()->validation_errors
- added curly brackets to if
- used BOOST_REQUIRE_THROW in sstable_compaction_test.cc
- moved injection to fill_buffer_from_underlying

Deexie commented:
- 1 is returned only for compaction_aborted_exception

bhalevy commented on compaction/compaction.hh:
```diff
@@ -85,12 +85,29 @@ struct compaction_data {
     uint64_t validation_errors = 0;
+
+    compaction_counters& operator+=(const compaction_counters& r) {
+        ended_at = r.ended_at;
```
better keep the maximum rather than just override the current `ended_at`

Deexie commented:
- maximum of ended at is kept

bhalevy reviewed (commented):
Looks very nice now.
We only need to resolve the util.py issue...

bhalevy commented:
Can be taken out of draft mode IMO

denesb commented on compaction/compaction.cc:
```diff
@@ -1242,12 +1245,16 @@ class scrub_compaction final : public regular_compaction {
             if (_scrub_mode == compaction_type_options::scrub::mode::abort) {
                 throw compaction_aborted_exception(_schema->ks_name(), _schema->cf_name(), "scrub compaction found invalid data");
             }
+            if (_scrub_mode == compaction_type_options::scrub::mode::segregate) {
```
I think the if is unnecessary, we can unconditionally bump the counter here.

denesb commented on compaction/compaction.cc:
```diff
@@ -1419,6 +1427,7 @@ class scrub_compaction final : public regular_compaction {
     std::string _scrub_start_description;
     mutable std::string _scrub_finish_description;
     uint64_t _bucket_count = 0;
+    compaction_task_stats_ptr _cts = seastar::make_lw_shared<compaction_manager::task::compaction_task_stats>();
```
Isn't a plain reference enough here?

Deexie replied:
It is a `lw_shared_ptr` because of `make_scrubbing_reader` function: https://github.com/scylladb/scylla/blob/19b689e44fcc48f7a86144dbaacbc8e42b8d53a9/compaction/compaction.cc#L1493-L1496
It is used in a test only, so changing that would not be painful, but requires playing with the lifetime of a variable the value of which is not even needed for the test (to avoid dangling reference). What do you think?

denesb replied:
I think it is worth the change, but like I said I'm not sure we even need this `compaction_task_stats` anymore. As far as I understand the initial purpose of this was to be an output parameter to `compact_sstables()`, but I see that we now do that via `compaction_result` which is a much better way of doing this IMO.

Deexie replied:
It's been changed to `uint64_t`. I will change that to reference.

denesb commented on compaction/compaction.cc:
```diff
@@ -1472,11 +1481,18 @@ class scrub_compaction final : public regular_compaction {
 
+    compaction_result finish(std::chrono::time_point<db_clock> started_at, std::chrono::time_point<db_clock> ended_at) override {
+        auto ret = compaction::finish(started_at, ended_at);
+        ret.validation_errors = _cts->validation_errors;
```
If we use `compaction_result` to communicate back validation error count, why do we need a `_cts` passed from outside? Can't we just have a local counter instead?

Deexie replied:
What do you mean by local?
We need to keep that as a member of `scrub_compaction` because during a single compaction many instances of reader are created with `make_sstable_reader()` and they need to share the same counter. I think it's clearer than passing a local variable through all the functions.

denesb replied:
I mean add a `uint64_t _validation_errors` member to `scrub_compaction` and we just pass reference to this member to scrub readers. Then in `finish()` we just do:
```
ret.validation_errors = _validation_errors;
```

denesb commented on compaction/compaction_manager.hh:
```diff
@@ -91,7 +91,19 @@ public:
+                tmp += r;
+                return tmp;
+            }
+        };
```
I think this struct is now redundant, supplanted by `compaction_result`.

denesb commented on test/boost/sstable_compaction_test.cc:
```diff
@@ -2383,8 +2383,8 @@ SEASTAR_THREAD_TEST_CASE(scrub_validate_mode_validate_reader_test) {
-        const auto valid = scrub_validate_mode_validate_reader(make_flat_mutation_reader_from_fragments(schema, permit, std::move(frags)), *info).get();
-        BOOST_REQUIRE(valid);
+        const auto errors = scrub_validate_mode_validate_reader(make_flat_mutation_reader_from_fragments(schema, permit, std::move(frags)), *info).get();
+        BOOST_REQUIRE(errors == 0);
```
`BOOST_REQUIRE_EQUAL(errors, 0);`

This way the value of `errors` will be printed if this fails.

denesb commented on test/boost/sstable_compaction_test.cc:
```diff
@@ -2394,8 +2394,8 @@ SEASTAR_THREAD_TEST_CASE(scrub_validate_mode_validate_reader_test) {
-        const auto valid = scrub_validate_mode_validate_reader(make_flat_mutation_reader_from_fragments(schema, permit, std::move(frags)), *info).get();
-        BOOST_REQUIRE(!valid);
+        const auto errors = scrub_validate_mode_validate_reader(make_flat_mutation_reader_from_fragments(schema, permit, std::move(frags)), *info).get();
+        BOOST_REQUIRE(errors != 0);
```
BOOST_REQUIRE_NE

denesb replied:
Use it also below.

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -1138,12 +1140,13 @@ class compaction_manager::rewrite_sstables_compaction_task : public compaction_m
             auto sst = consume_sstable();
-            co_await rewrite_sstable(std::move(sst));
+            const auto& res = co_await rewrite_sstable(std::move(sst));
+            // FIXME: utilize compaction result
```
What does this refer to? I see it is removed in the next commit.

Deexie replied:
To the `res` that could be discarded in this commit. Is it not needed here?

denesb replied:
I think the above change is unnecessary. You can just keep the previous `co_await rewrite_sstable(std::move(sst));` as-is and only save the result into a variable in the commit in which you actually use it. It is fine to discard return-values in most cases.

denesb commented on compaction/compaction.hh:
```diff
@@ -85,12 +85,29 @@ struct compaction_data {
 
-struct compaction_result {
-    std::vector<sstables::shared_sstable> new_sstables;
+struct compaction_counters {
```
We usually call these kind of structs `stats`, I think `compaction_stats` would be a better name.

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -412,15 +419,15 @@ class compaction_manager::custom_compaction_task : public compaction_manager::ta
 
 protected:
-    virtual future<> do_run() override {
+    virtual future<compaction_counters_opt> do_run() override {
```
Why use optional here? Can't we just return default constructed stats struct instead?

denesb replied:
In any case this (adding the return value to `perform_task()` should be in a commit of its own.

Deexie replied:
At first it returned default constructed but scrub specific struct and the change to optional was requested. Currently it is general for all compaction tasks though. Should it be changed to default constructed value? @bhalevy @denesb

denesb replied:
I don't have a strong opinion on this.

denesb replied:
Actually: with the current interface, one has to know which concrete task actually returns an engaged optional. With default-initialized stats struct, one can just use the counters returned, they will be 0 if not set. @bhalevy WDYT?

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -1098,16 +1109,18 @@ class compaction_manager::offstrategy_compaction_task : public compaction_manage
             }
         }
+
+        co_return std::nullopt;
```
I very much dislike this return value added for a single concrete task's sake. Not sure what would be the better alternative though.

denesb commented on api/storage_service.cc:
```diff
@@ -1403,16 +1411,28 @@ void set_snapshot(http_context& ctx, routes& r, sharded<db::snapshot_ctl>& snap_
+            counters += lhs.value();
+            counters += rhs.value();
+            return counters;
+        };
```
You can just add `operator+()` to `compaction_counters`, then use `std::plus<compaction_counters>` in the `map_reduce` below.

denesb replied:
Oh, it is an optional. This would work OOTB if we used plain values instead of optional.

denesb commented on api/storage_service.cc:
```diff
@@ -1427,8 +1427,15 @@ void set_snapshot(http_context& ctx, routes& r, sharded<db::snapshot_ctl>& snap_
+        }).then_wrapped([] (auto f) {
+            if (f.failed()) {
+                auto ex = f.get_exception();
+                if (try_catch<sstables::compaction_aborted_exception>(ex)) {
```
Can't we add this to the return value of `perform_task()` too? So it returns a status bit (success, abort, error, etc).

Deexie replied:
I am not sure how status bit could be reduced.

Deexie replied:
Do you mean `perform_task()` should return status bit together with `compaction_stats` or just a status bit?
For the first version it could be something like `std::pair<compaction_stats, compaction_status>` and then mapping it to the `first` member?

denesb replied:
I thought about something like this:
```
struct compaction_result { // name is debatable
   status status;
   compaction_stats stats;
};
```
We avoid using `std::pair<>` in public APIs, as it is ugly and using a struct is an opportunity to document the purpose with a good name.

Deexie replied:
Currently `perform_task()` throws an exception for aborted operation. Changing it so that it returns `status::aborted` would require changes in behaviour of
```
perform_sstable_scrub_validate_mode()
perform_task_on_all_files()
perform_offstrategy()
compaction_manager::submit()
compaction_manager::run_custom_job()
perform_major_compaction()
```
And all of their callers. Should I really do that? I guess rethrowing or changing of rest api would be needed anyway.

We could also check the retruned value of `perform_task()` in the four latest function and same for `perform_cleanup()` that calls `perform_task_on_all_files()` and if compaction was aborted then rethrow the exception. It doesn't sound reasonable, though.

We could also pass an additional bool parameter set to false on default to `perform_task()` and basing on it throw an exception or just mark it in the status code. That would only obfuscate the code I think.

Or maybe you were thinking about some other solution?

denesb replied:
Indeed, that is a lot of churn. I still think this would be cleaner, but let's do it as a follow-up.

Deexie commented:
- unconditional counter bump
- `BOOST_REQUIRE` -> `BOOST_REQUIRE_EQUAL`/`BOOST_REQUIRE_NE`
- got rid of `compaction_task_stats`
- compaction_counters -> compaction_stats
- change of return type of perform_task moved to a distinct patch

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1545/

denesb reviewed (changes requested):
Mostly good. I had just two material comments:
* Change `scrub_compaction::_validation_errors` to be just a plain `uint64_`.
* Use a status bit to communicate result of scrub to API layer, instead of an exception.

And one cosmetic comment about moving a hunk between 2 patches.

Deexie commented:
- moved scylla_inject_error to cql-pytest/util.py
- rebased on current master

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1550/

Deexie commented:
- `_validation_errors` change from `lw_shared_ptr<uint64_t>` to `uint64_t`
- discard `rewrite_sstable`/`validate_sstable` result and got rid of FIXME

denesb commented on compaction/compaction.hh:
```diff
@@ -85,12 +85,29 @@ struct compaction_data {
+
+struct compaction_result {
+    std::vector<sstables::shared_sstable> new_sstables;
+    compaction_stats counters;
```
nit: also rename member to `stats`

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -285,7 +287,7 @@ future<> compaction_manager::perform_task(shared_ptr<compaction_manager::task> t
 
     try {
-        co_await task->run();
+        auto&& res = co_await task->run();
```
nit: unused variable `res`

denesb replied:
if a later commit uses it, add it in said commit

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -12,8 +12,10 @@
+#include <optional>
 #include <seastar/core/metrics.hh>
 #include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
```
nit: I doubt this and the <optional> include was really needed here, they are probably included indirectly already. Only add new includes if the file doesn't compile.

denesb reviewed (approved):
LGTM, nits only.
@bhalevy please re-review.

Deexie commented:
- counters -> stats
- deleted unused variable
- deleted unnecessary imports

bhalevy commented on test/alternator/conftest.py:
```diff
@@ -13,7 +13,7 @@
 import requests
 import re
-from util import create_test_table, is_aws
+from alternator_util import create_test_table, is_aws
```
@nyh please review and ack this patch

bhalevy commented on test/alternator/test_batch.py:
```diff
@@ -8,8 +8,12 @@
+from alternator_util import random_string, full_scan, full_query, multiset
+
+sys.path.insert(1, sys.path[0] + '/../cql-pytest')
+from util import scylla_inject_error
```
yeah, that makes sense until we have a more formal way to provide a common library to all our scylla pytests

bhalevy commented on test/rest_api/test_storage_service.py:
```diff
@@ -10,7 +10,7 @@
 # Use the util.py library from ../cql-pytest:
 sys.path.insert(1, sys.path[0] + '/../cql-pytest')
-from util import unique_name, new_test_table, new_test_keyspace, new_materialized_view, new_secondary_index
+from util import unique_name, new_test_table, new_test_keyspace, new_materialized_view, new_secondary_index, scylla_inject_error
```
This hunk probably belongs to the next patch.
It has no use in this one.

bhalevy reviewed (changes requested):
Looks good.  I had some small comments
and asked @nyh's ack on the alternator/util.py change.

Deexie commented:
- auto& -> auto
- counters -> statistics in commit message and counters -> stats in code
- moved import to a proper commit

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1570/

denesb commented:
@Deexie please investigate why CI continuously fails.

denesb commented:
> @Deexie please investigate why CI continuously fails.

Nevermind, last CI failed because of the "six" fiasko. I restarted it.

bhalevy commented:
> > @Deexie please investigate why CI continuously fails.
> 
> Nevermind, last CI failed because of the "six" fiasko. I restarted it.

The branch needs to be rebased to get the fix though...

Deexie commented:
- rebased on master

psarna commented:
I haven't seen any CI update since then, I think it was due to the temporary failure following our scylladb/scylla -> scylladb/scylladb rename. Retriggered

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1606/
